### ISSUE TITLE
Azure: Dapr fails to obtain the authentication token when using Managed Identity of Azure Web App

### ISSUE THREAD

**Issue by jogobella00** [dapr/components-contrib#2913]
## In what area(s)?

> /area runtime

## What version of Dapr?

> 1.11.0

## Expected Behavior

On startup of the sidecar Dapr will get the authentication token using System Assigned Managed Identity of the Azure Web App and generate components that will communicate with Azure services.

## Actual Behavior

Fails to obtain the auth token:

`level=fatal msg="process component binding-azure-storage-my-component error: [INIT_COMPONENT_FAILURE]: initialization error occurred for binding-azure-storage-my-component (bindings.azure.blobstorage/v1): failed to create Azure Storage container myContainer: ChainedTokenCredential: failed to acquire a token.\nAttempted credentials:\n\tmanaged identity\n\tAzureCLICredential: Azure CLI not found on path" app_id="$Default" instance=<redacted> scope=dapr.runtime type=log ver=1.11.0`

After encountering this error I tried to have workaround by installing the Azure CLI in the docker image of the Azure Web App, but then Dapr times out when creating the component:

`level=fatal msg="process component biding-azure-storage-my-component error: [INIT_COMPONENT_FAILURE]: initialization error occurred for biding-azure-storage-my-component (bindings.azure.blobstorage/v1): init timeout for component biding-azure-storage-my-component exceeded after 5s" app_id="$Default" instance=<redacted> scope=dapr.runtime type=log ver=1.11.0`

## Steps to Reproduce the Problem

Create an Azure Web App with System Assigned Managed Identity.
Build a docker image with your app and Dapr.

Inject Dapr component to the Web App, make sure that this component will use System Assigned Managed Identities to authenticate in Azure e.g.:
```yaml 
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: binding-azure-storage-my-component
  namespace: default
spec:
  type: bindings.azure.blobstorage
  version: v1
  metadata:
  - name: storageAccount
    value: <storageAccountName>
  - name: container
    value: <containerName>
```

Start the application.

*Issue is __not__ existing in the previous Dapr version (1.10.7).

## Release Note

RELEASE NOTE: **FIX** Dapr to obtain authentication token when using Managed Identity of Azure Web App.

**Comment by berndverst** [comment]
Maybe the timeout needs to be increased here: https://github.com/dapr/components-contrib/blob/master/internal/authentication/azure/auth.go

Or we have to look into how to proceed even if the Azure CLI returns an error.

This only impacts *slow* environments where the MSI endpoint responds slowly.

If the CLI is not available it should actually retry MSI again. I'll take a look at why that isn't occurring. May need some help from the Azure SDK team here as I am largely limited by their SDK.

**Comment by XavierGeerinck** [comment]
I am not 100% sure as this started happening since 1.11.0 and was fine in 1.10.7. Did something change at the MSI side of things making it respond more slowly now?

**Comment by berndverst** [comment]
> I am not 100% sure as this started happening since 1.11.0 and was fine in 1.10.7. Did something change at the MSI side of things making it respond more slowly now?

In 1.11 we added support for Workload Identity and Azure CLI Credential. As part of this we standardized our implementation to match the DefaultAzureCredential implementation from the official Azure SDK for Go / Azure Identity SDK.

In that implementation a default Managed Identity time out of 1 second is used so that quick fall-through to the CLI credential is possible. However, it appears that in App Service 1 second is insufficient.

Removing the timeout completely breaks the use of the CLI credential has the MSI attempt would hang indefinitely.

Of course increasing the timeout is a bad experience for local development, and arbitrary running in App Service too.

I have now come up with a solution that disables the timeout only in an environment which appears to be using certain managed identity mechanisms.

**Comment by berndverst** [comment]
For reference, here is the managed identity client code from the default azure credential in the official Azure SDK for Go:
https://github.com/Azure/azure-sdk-for-go/blob/fb99a64943cb1d4594ce1e1bc0c0feeb90d083a8/sdk/azidentity/default_azure_credential.go#L102-L108

**Comment by ItalyPaleAle** [comment]
Fixed in the release-1.11 branch by #2916
